packages/core/src/utils/parse-type-id.ts

```diff
@@ -0,0 +1,104 @@
+import assert from 'assert';
+
+// This module parses type identifiers.
+
+export interface ParsedTypeId {
+  id: string;
+  head: string;
```

Comment:
Do you think we could type `head`, so we make sure that any checks that depend on this can be made exhaustive?

Reply:
I prefer to keep `string` because Solidity could add new types in the future. Any sort of exhaustive check would be incorrect in that scenario, so it's better to force handling unknown types. In `compatibleTypes` any unknown types are conservatively considered incompatible.